PR: https://github.com/odoo/odoo/pull/119320

From: bbfd5aa614d49e41873d70fee9758cd059dbfbee
From: Surabhi Varma
Date: 2023-07-03 10:20:58

Structural Changes: 5
Total Changes: 20

[IMP] stock_delivery: route based on shipping method

In this commit, we added a new field in the shipping methods this field can be
used select rules. These rules can only be used when the corresponding SO has
one of the selected shipping methods.

task-3236149

closes odoo/odoo#119320

Signed-off-by: Arnold Moyaux (arm) <arm@odoo.com>

================================= pseudo patch: =================================

--- a/addons/stock_delivery/models/delivery_carrier.py
+++ b/addons/stock_delivery/models/delivery_carrier.py
@@ -24,6 +24,10 @@ class DeliveryCarrier(models.Model):
         "shipping will be updated on the SO after the delivery."
     )
 
+    route_ids = fields.Many2many(
+        'stock.route', 'stock_route_shipping', 'shipping_id', 'route_id', 'Routes',
+        domain=[('shipping_selectable', '=', True)])
+
     # -------------------------- #
     # API for external providers #
     # -------------------------- #

--- a/addons/stock_delivery/models/sale_order.py
+++ b/addons/stock_delivery/models/sale_order.py
@@ -42,3 +42,13 @@ class SaleOrder(models.Model):
         else:
             post = u'\N{NO-BREAK SPACE}{symbol}'.format(symbol=self.currency_id.symbol or '')
         return u' {pre}{0}{post}'.format(amount, pre=pre, post=post)
+
+
+class SaleOrderLine(models.Model):
+    _inherit = 'sale.order.line'
+
+    def _prepare_procurement_values(self, group_id):
+        values = super(SaleOrderLine, self)._prepare_procurement_values(group_id)
+        if not values.get("route_ids") and self.order_id.carrier_id.route_ids:
+            values['route_ids'] = self.order_id.carrier_id.route_ids
+        return values

--- a/addons/stock_delivery/models/stock_move.py
+++ b/addons/stock_delivery/models/stock_move.py
@@ -4,6 +4,12 @@ from odoo import api, fields, models
 from odoo.tools.sql import column_exists, create_column
 
 
+class StockRoute(models.Model):
+    _inherit = "stock.route"
+
+    shipping_selectable = fields.Boolean("Applicable on Shipping Methods")
+
+
 class StockMove(models.Model):
     _inherit = 'stock.move'
 
